mithril-common/src/chain_observer/pallas_observer.rs

```diff
@@ -78,6 +83,81 @@ impl PallasChainObserver {
         Ok(epoch)
     }
 
+    /// Returns inline datum bytes from the given `Values` instance.
+    fn get_datum_bytes(&self, utxo: &Values) -> Vec<u8> {
+        let bytes = utxo.inline_datum.as_ref().unwrap().1.clone();
+        let bytes = CborWrap(bytes).to_vec();
+        inspect(bytes)
+    }
+
+    /// Returns inline datums from the given `Values` instance.
+    fn inspect_datum(&self, utxo: &Values) -> Datum {
+        let datum = self.get_datum_bytes(utxo);
+
+        inspect(datum)
+    }
+
+    /// Serializes datum to `TxDatum` instance.
+    fn serialize_datum(&self, utxo: &Values) -> TxDatum {
+        let datum = self.inspect_datum(utxo);
+        let serialized = serde_json::to_string(&datum.to_json()).expect("Failed to serialize");
+
+        TxDatum(serialized)
+    }
+
+    /// Maps the given `UTxOByAddress` instance to Datums.
+    fn map_datums(&self, transaction: UTxOByAddress) -> Datums {
+        transaction
+            .utxo
+            .iter()
+            .filter(|(_, utxo)| utxo.inline_datum.is_some())
+            .map(|(_, utxo)| self.serialize_datum(utxo))
+            .collect()
```

Comment:
You could have probably used `FilterMap` here:
```suggestion
        transaction
            .utxo
            .iter()
            .filter_map(|(_, utxo)| utxo.inline_datum.map(|d| self.serialize_datum(d)))
            .collect()
```